tests/serac_dtor.cpp

```diff
@@ -32,59 +27,57 @@ TEST(serac_dtor, test1)
 
   // Open the mesh
   ASSERT_TRUE(file_exists(mesh_file));
+
   std::ifstream imesh(mesh_file);
-  mfem::Mesh *  mesh = new mfem::Mesh(imesh, 1, 1, true);
+
+  auto mesh = std::make_unique<mfem::Mesh>(imesh, 1, 1, true);
+
   imesh.close();
 
   // Refine in serial
   mesh->UniformRefinement();
 
-  // Declare pointer to parallel mesh object
-  mfem::ParMesh *pmesh = nullptr;
-
   // Initialize the parallel mesh and delete the serial mesh
-  pmesh = new mfem::ParMesh(MPI_COMM_WORLD, *mesh);
-  delete mesh;
+  auto pmesh = std::make_shared<mfem::ParMesh>(MPI_COMM_WORLD, *mesh);
 
   // Refine the parallel mesh
   pmesh->UniformRefinement();
 
   // Initialize the second order thermal solver on the parallel mesh
-  ThermalSolver therm_solver(2, pmesh);
+  auto therm_solver = std::make_unique<ThermalSolver>(2, pmesh);
 
   // Set the time integration method
-  therm_solver.SetTimestepper(TimestepMethod::QuasiStatic);
+  therm_solver->SetTimestepper(TimestepMethod::QuasiStatic);
 
   // Initialize the temperature boundary condition
-  mfem::FunctionCoefficient u_0([](const mfem::Vector &x) { return x.Norml2(); });
+  auto u_0 = std::make_shared<mfem::FunctionCoefficient>([](const mfem::Vector &x) { return x.Norml2(); });
 
   std::vector<int> temp_bdr(pmesh->bdr_attributes.Max(), 1);
 
   // Set the temperature BC in the thermal solver
-  therm_solver.SetTemperatureBCs(temp_bdr, &u_0);
+  therm_solver->SetTemperatureBCs(temp_bdr, u_0);
 
   // Set the conductivity of the thermal operator
-  mfem::ConstantCoefficient kappa(0.5);
-  therm_solver.SetConductivity(kappa);
+  auto kappa = std::make_shared<mfem::ConstantCoefficient>(0.5);
+  therm_solver->SetConductivity(kappa);
 
   // Define the linear solver params
   LinearSolverParameters params;
   params.rel_tol     = 1.0e-6;
   params.abs_tol     = 1.0e-12;
   params.print_level = 0;
   params.max_iter    = 100;
-  therm_solver.SetLinearSolverParameters(params);
+  therm_solver->SetLinearSolverParameters(params);
 
   // Complete the setup without allocating the mass matrices and dynamic
   // operator
-  therm_solver.CompleteSetup();
-
-  // just do something to make sure the dtor is being called
-  // and that the member variables lifetime is managed properly
-  do_nothing(therm_solver);
-  do_nothing(therm_solver);
-  do_nothing(therm_solver);
-  do_nothing(therm_solver);
+  therm_solver->CompleteSetup();
+
+  // Destruct the old thermal solver and build a new one
+  therm_solver.reset(new ThermalSolver(1, pmesh));
+
+  // Destruct the second thermal solver and leave the pointer empty
+  therm_solver.reset(nullptr);
```

Comment:
hmm, I wonder if it even makes sense to keep this test around if the solvers have trivial deconstructors.

Reply:
I think it's a good idea to keep it around, especially once we get memory checking going. I can see people adding more complicated stuff, and this is a good check for it.